### DISCUSSION TITLE
5월 2주 스터디 일지

### DISCUSSION THREAD

**Discussion by brightyein** [KKLP01/study#28] [Announcements]
## 스터디 진행 방식

1. 각자 과제하면서 어려움을 느꼈던 부분과 어떻게 해결했는지 중점으로 이야기
2. 다른 사람들 소스코드에서 얻은 인사이트 내껄로 만들기 (각자 코드에 적용해보기)
3. 다른 사람 pr에다가 코드리뷰 다는 습관 기르기

## 과제 출제 방식

1. 취약한 부분을 중점으로 과제 만들어서 진행하기
2. 배운 부분을 과제만들어서 진행하기

## 데일리 스터디 과제

월요일_아이디 중복체크 기능 > 화요일 출근 전까지

**Comment by brightyein** [comment]
### <잘못알고 있었던 부분>

객체의 배열을 아래와 같이 생성하면,
private final static User[] users = new User[10];

아래와 같은 의미라고 생각했음.
private final static User[] users = new User[]{new User(),new User(),new User(),new User(),new User(),new User(),new User(),new User(),new User(),new User()};

그런데, 아래와 같은 의미였음.
private final static User[] users = new User[]{null,null,null,null,null,null,null,null,null,null};

---
### <헷갈렸던 부분>
- 상황에 따라 for 문을 종료시키는 코드 

continue : for 문이 순회 할 여러 인덱스 중 현재의 인덱스만 건너뛴다
return : 그 뒤 인덱스도 아예 실행하지 않고 즉시 for 문을 끝낸뒤 그 함수를 종료시키고 값을 반환한다
break : 반복을 완전히 종료하고 for 문을 빠져나간다 / 반복문 속의 반복문이라면 바로 지금의 반복을 끝낸다

---
### <앞으로 코딩하며 익숙해지며 좋은 습관>
- 깃은 인텔리제이 GUI 로 사용하기
- 단축키에 익숙해질 것 (그래야 속도가 빨라진다)
- 완료 후 지우더라도 진행과정을 로그로 콘솔에 찍어보면서 할 것
- 디버그 기능을 사용할 것

**Reply by kimdoyeon304** [reply]
단축키 인정!

**Comment by kimdoyeon304** [comment]
1. users[i].equals(newUser)
users[i].getId().equals(newUser.getId())&&users[i].getPwd().equals(newUser.getPwd())

객체배열의 내용은 객체이니 객체끼리 동일여부가능할줄 알았는데 아니었음
배열안에 배열, 2차원배열처럼 생각하여 처리함


2. users[i] = " ";
users[i] = null;

배열 초기화를 " "으로 하는 경우도 있어, 가능할 줄 알았음
배열의 기본초기값으로 초기화


3. 입력란이 연이어 있을때, 두개가 동시에 뜸
제대로 입력하기 어려움

nextInt(), next() 등 이어 nextLine()이면 버퍼가 생길 수 있음
nextLine() 을 그사이에 선언하여 버퍼처리


4. 향상된 for문(foreach문)
for문을 생각못하고 맨상단에 위치->어떤 작업 후 메인으로 안가고 해당 메소드 맨상단으로 감

로직흐름파악해서 필요한 부분에 재위치


등등... 객체배열, 객체부터 접근제한자, 메소드 매개변수, 리턴타입
그리고 전체적인 흐름은 파악했으나 적재적소에 코드구현하는 것...이 부족...합니다.........

**Reply by brightyein** [reply]
4번이 텍스트로만 보면 무슨 말인지 이해가 잘 안가요😭
코드 덧붙여 주실 수 있나요오? 아니면 어느 부분인지 깃헙 링크로 달아주시면 좋을 듯 합니다:)

마지막.... 너무 공감합니다!

**Reply by kimdoyeon304** [reply]
`private final static Member[] members;

public static Member[] findAllMembers() {
        return members;
    }

-------------------------------------------------------------------------------

`Scanner sc = new Scanner(System.in);
    LoginRepository loginRepository = new LoginRepository();

    public void login() {

        for (User user : loginRepository.findAllUsers()) {
            if (user == null) {
                System.out.println("가입한 회원이 없습니다");
                System.out.println("회원가입 창으로 이동합니다");
                signUpUsers();
            } else {
                while (true) {
                    System.out.println("아이디 입력");
                    String id = sc.nextLine();
                    System.out.println("비밀번호 입력");
                    String pwd = sc.nextLine();

                    User newUser = new User(id, pwd);

                    if (newUser.getId().equals(user.getId()) && newUser.getPwd().equals(user.getPwd())) {
                        System.out.println("---------로그인 완료------");
                        break;
                    } else {
                        System.out.println("------로그인 실패-------");
                        System.out.println("아이디나 비밀번호가 다릅니다. 다시 입력하세요");
                    }
                }
                System.out.println("---------프로그램 종료------");
                System.exit(0);
            }
        }
    }`

선생님과 함께한 실습 중에 향상된 for문 쓰임보고 문제 풀때 적용해서 풀었는데
별 생각없이 맨 상단에 지정하고 작업을 했더니, 프로그램 종료문장 System.exit(0); 없이 break만 썼을때,
로그인 완료 후 메인으로 가길 바랐으나 계속 반복문 while문만 나가서 for문으로 가 다시 내용을 수행하더라구요ㅠ
그때는 뭐가 문제인지 잘 모르고 메인으로 가기위해 break에다가  System.exit(0);까지 적용했어요....

객체배열 속 값을 하나씩 꺼내 객체로 받아 for문을 도는걸 생각했어야 했어요 ㅎㅎ..
값 수에 따라 도는거죠...
----------------------------------------------------------------------------
`public void login() {           //로그인

        System.out.println("====================");
        if (loginRepository.findAllUsers()[0] == null) {            //DB확인
            System.out.println("가입한 회원이 없습니다");
            System.out.println("회원가입 창으로 이동합니다");
            signUpUsers();          //회원가입
        } else {
            while (true) {
                System.out.println("아이디 입력");
                String id = sc.nextLine();
                System.out.println("비밀번호 입력");
                String pwd = sc.nextLine();

                User newUser = new User(id, pwd);
                //향상된 for문(foreach문)
                for (User user : loginRepository.findAllUsers()) {
                    if(user!=null) {            //DB확인
                        if (newUser.getId().equals(user.getId()) && newUser.getPwd().equals(user.getPwd())) {
                            System.out.println("---------로그인 완료------");
                            loginOn(newUser);           //로그인 후 메인
                            return;         //메인 홈
                        }
                    }
                }
                System.out.println("------로그인 실패-------");
                System.out.println("아이디나 비밀번호가 다릅니다. 다시 입력하세요");
            }
        }
    }`

이거는 이번에 회원탈퇴기능 추가하면서 수정한 부분입니다
해당 구문을 맨상단에서 필요한 부분에 넣어 원하는 곳만 조건에 맞춰 돌고 나갈 수 있도록 했습니다

이해가 가실까요? ㅠㅠ

**Reply by brightyein** [reply]
이해했습니다!! 굿굿

**Comment by YJ-2** [comment]
- 객체배열을 삭제하는 부분
특정 값만 null 처리를 하게 되면 중간이 비워진 상태가 되므로
기존 값들을 한 칸씩 앞당겨서 특정 값이 삭제되도록 처리함

- 객체 지향 프로그래밍
어떤 식으로 작성해야 객체 지향 프로그래밍인지,
클래스와 메소드를 어떤 기준으로 나눠야 할 지 등 
연습해야함

**Reply by kimdoyeon304** [reply]
2번 대공감ㅠ

**Comment by eunahpae** [comment]
- 단일 책임 원칙으로 메소드들을 다 쪼개다보니, 여기저기서 호출해서 쓰는게 너무 헷갈리는 것 같음. 
- 개발자적 사고가 안되서, 코드를 구상하기가 쉽지 않음. 코드를 작성할 순서를 순차적으로 먼저 정리한 후 코드를 작성해야 할 것 같음.

**Reply by kimdoyeon304** [reply]
순서정하기 어렵죠...ㅎㅏㅠ

**Reply by brightyein** [reply]
저도 service 와 repository, dto 등이 어떤 흐름으로 연결되는지 한번에 사고가 안되더라구요 ㅠㅠ
연습만이 살길이다!!
